packages/SwingSet/src/kernel/kernel.js

```diff
@@ -305,25 +303,36 @@ export default function buildKernel(
     }
   }
 
-  function resolveToError(kpid, errorData, expectedDecider) {
-    insistCapData(errorData);
-    const p = kernelKeeper.getResolveablePromise(kpid, expectedDecider);
-    const { subscribers, queue } = p;
-    let idx = 0;
-    for (const dataSlot of errorData.slots) {
-      kernelKeeper.incrementRefCount(dataSlot, `reject|s${idx}`);
-      idx += 1;
-    }
-    kernelKeeper.rejectKernelPromise(kpid, errorData);
-    // we don't notify the decider; they already know
-    notifySubscribersAndQueue(kpid, expectedDecider, subscribers, queue);
-    if (p.policy === 'logFailure' || p.policy === 'logAlways') {
-      console.log(
-        `${kpid}.policy ${p.policy}: rejection ${JSON.stringify(errorData)}`,
-      );
-    } else if (p.policy === 'panic') {
-      panic(`${kpid}.policy panic: rejection ${JSON.stringify(errorData)}`);
+  function doResolve(vatID, resolutions) {
+    if (vatID) {
```

Comment:
Something feels a bit funny about having `vatID` be optional. Not necessarily wrong, but I'm still walking through the callers to understand what it gets used for and who provides it (and who doesn't).

I think this is basically the decider of the batch of promises being resolved. When passed to `kernelKeeper.getResolveablePromise()`, it's used to double-check the recorded decider of the given promise. In that sense, `vatID` must exactly match: if the promise is being decided by the kernel, `vatID` must be `undefined` (it's not an optional parameter.. it's a required parameter whose value is sometimes `undefined` because that's what we used to mean "the kernel is the decider", which I'll admit is maybe not the best choice of representation, but without something like Rust's `enum` it's the most convenient we've got).

In the call to `notifySubscribersAndQueue`, the `vatID` is used to refrain from queueing a notification to the resolving vat (if they emitted the resolution, they should not be told about the resolution later). In this case, if `vatID` is `undefined`, then a notification will be queued for every subscriber.

`doResolve` is called from a syscall handler (with `vatID` of the vat doing the syscall, who is clearly doing the resolution), and from within the kernel in `resolveToError`. `resolveToError` is called from:

* When the target vat is being terminated, any orphaned promises it was responsible for (i.e. it was their decider) are rejected. This uses the late vat's `vatID` in the call to `doResolve`, which is correct for both the who-is-the-decider check and the don't-notify-the-resolving-vat check.
* After the vat is terminated, subsequent messages to that vat (in `deliverToVat`) have their `msg.result` rejected with VAT_TERMINATION_ERROR, *without* a `vatID`. The kernel is still the decider for these (up until the message is translated into vatspace), so it is appropriate for `doResolve` to use `vatID = undefined`.
* As messages reach the top of the run-queue, in `deliverToTarget`, we might find that:
  * the target is an orphaned object (the "owner" field is empty), which triggers `resolveToError` without a vatID, which is correct again because the kernel is the decider for the result.
  * the target is a promise which is fulfilled to a non-Presence, which means the kernel is responsible for rejecting any result promise with a "no method" error. The result promise is decided by the kernel at that point, so `vatID=undefined` is correct
  * the target is a promise which is in a rejected state, so any result promise must be rejected with exactly the rejection data. That data might contain additional promises, but I think that doesn't matter. Using `vatID=undefined` remains correct for the same resonses as above.
  * the target is an unresolved promise, whose decider is some other vat, but that vat has been terminated (and we somehow didn't clear it out earlier). Any result promise gets rejected with the usual VAT_TERMINATION_ERROR, and `vatID=undefined` is correct because the result promise is still decided by the kernel at that point.

Ok, so I've convinced myself that `vatID=undefined` is a legitimate and necessary situation.

Reply:
The pattern of using a `vatID` of `undefined` to mean "the resolver is not a vat, it's the kernel" struck me as surprising when I stumbled across this pattern while making this set of changes.  My first thought was to change the code to use `'kernel'` as the "vat ID" instead of undefined in order to be explicit about what's going on, but when I tried that it started unfolding into a bunch of other obnoxious changes so I backed off and returned to using undefined to signal that.  Like you, I thought this pattern was a bit goofy when I first encountered it, but it's been this way since forever and I think it's OK.